### DISCUSSION TITLE
assignByRef is gone, long live... who?

### DISCUSSION THREAD

**Discussion by SinusPi** [smarty-php/smarty#1081] [General]
I used to abuse the assignByRef function to create a way for templates to "talk back" to the calling code, not merely returning a block of rendered HTML, but a few variable values, too - like giving the templates a way to return a tuple instead of a plain string, if they were functions.
So with assignByRef gone in new Smarty versions, what would be the best approach to let templates return "something more"?
I'm almost set on writing some simple plugin that would store values "set" by the template, but I wonder if there's a better idea.

**Comment by wisskid** [comment]
Hi @SinusPi! Please see https://smarty-php.github.io/smarty/stable/designers/language-builtin-functions/language-function-assign/#examples for some examples on how to use `$smarty->getTemplateVars()` to read variables set in templates (e.g. using `{assign}`).

**Reply by SinusPi** [reply]
So, to answer my own question more directly, using `$t->assign("bucket",$BUCKET); $t->fetch(); /* {$BUCKET.foo="bar"} */ $BUCKET=getTemplateVars('bucket');` it's possible to emulate the old `$t->assignByRef("bucket",$BUCKET); $t->fetch();` behaviour, am I correct?

**Reply by wisskid** [reply]
Correct! 👍

**Reply by SinusPi** [reply]
It helps - somewhat. But, what would I do in a scenario when I no longer have access to `$t` after it's displayed, because it's part of a buffered `include('i_display_a_template.php');` call? With refs, assigned variables would still be updated. Would abusing an `output`-type plugin work in this case, doing the `getTemplateVars` and exporting to globals/wherever?

**Reply by wisskid** [reply]
Yes, I guess that would be possible.

**Comment by ssigwart** [comment]
@SinusPi, I submitted #986 to bring back support for `assignByRef`. You could create a fork and see if that works for you.
